src/UiPath.Workflow.Runtime/Statements/Flowchart.cs

```diff
@@ -179,199 +43,166 @@ protected override void CacheMetadata(NativeActivityMetadata metadata)
-        }
+            Fx.Assert(node.Index == -1 && !_reachableNodes.Contains(node), "Corrupt Flowchart.reachableNodes.");
+            node.Index = _reachableNodes.Count;
+            _reachableNodes.Add(node);
+            node.CacheMetadata(this, metadata);
 
+            return node.GetSuccessors();
+        }
 
-        Fx.Assert(node != null, "caller should validate");
-        FlowNode current = node;
-        do
+        void PropagateSplitsStack(FlowNode predecessor, FlowNode successor)
         {
-            if (ExecuteSingleNode(context, current, out FlowNode next))
-            {
-                current = next;
-            }
-            else
+            if (predecessor is null || successor is null)
+                return;
+            var predecessorStack = GetStaticSplitsStack(predecessor);
+            var successorStack = GetStaticSplitsStack(successor);
+            switch (predecessor)
             {
-                _currentNode.Set(context, current.Index);
-                current = null;
+                case FlowSplit split:
+                    successorStack.Push(split, predecessorStack);
+                    break;
+                case FlowMerge _:
+                    successorStack.AddPop(predecessorStack);
+                    break;
+                default:
+                    successorStack.PropagateStack(predecessorStack);
+                    break;
             }
         }
-        while (current != null);
     }
 
-    private bool ExecuteSingleNode(NativeActivityContext context, FlowNode node, out FlowNode nextNode)
+    private void ValidateSplitsAndMerges(NativeActivityMetadata metadata, FlowNode node)
     {
-        Fx.Assert(node != null, "caller should validate");
-        if (node is FlowStep step)
+        ValidateSingleSplitInAmonte(node);
+        switch (node)
         {
-            _onStepCompleted ??= new CompletionCallback(OnStepCompleted);
-            return step.Execute(context, _onStepCompleted, out nextNode);
+            case FlowMerge merge: ValidateMerge(merge); break;
+            case FlowSplit split: ValidateSplit(split); break;
         }
 
-        nextNode = null;
-        if (node is FlowDecision decision)
+        void ValidateSingleSplitInAmonte(FlowNode node)
         {
-            _onDecisionCompleted ??= new CompletionCallback<bool>(OnDecisionCompleted);
-            return decision.Execute(context, _onDecisionCompleted);
+            var splits = GetStaticSplitsStack(node).GetTop();
+            if (splits.Count > 1)
+                AddValidationError($"Node has multiple splits incoming branches. Please precede with a Merge node.", node, splits);
         }
+        void ValidateSplit(FlowSplit split)
+        {
+            HashSet<FlowMerge> allMerges = new();
+            foreach (var branch in split.Branches)
+            {
+                var merges = GetMerges(branch).Distinct().ToList();
+                allMerges.AddRange(merges);
+            }
```

Comment:
You can simplify by getting all the merges for a split, e.g.:
`var allMerges = GetMerges(split)`

and GetMerges can be:

`
List<FlowMerge> GetMerges(FlowSplit flowSplit)
{
    return (
        from nodeInfo in _staticBranchesByNode
        where nodeInfo.Key is FlowMerge
        where nodeInfo.Value.GetTop().Contains(flowSplit)
        select nodeInfo.Key as FlowMerge
        ).ToList();
}
`